filechooser-portal/Main.vala

```diff
@@ -343,10 +333,59 @@ public class Files.FileChooserPortal : Object {
         yield;
 
         dialogs.remove (parent_window);
+        dialog.destroy ();
+
         response = _response;
         results = _results;
     }
 
+    private async bool check_overwrite_uri (string? uri, Gtk.Window parent) {
+        if (uri == null) {
+            return false;
+        }
+
+        var file = GLib.File.new_for_uri (uri);
+        if (file.query_exists ()) {
+            var display_name = file.get_basename ();
+            var primary = _("Replace “%s”?").printf (display_name);
+            unowned var secondary = _("Replacing this file will overwrite its current contents");
+            if (file.query_file_type (FileQueryInfoFlags.NOFOLLOW_SYMLINKS) == FileType.SYMBOLIC_LINK) {
+                try {
+                    var info = file.query_info (FileAttribute.STANDARD_SYMLINK_TARGET, FileQueryInfoFlags.NONE);
+                    primary = _("Replace “%s”?").printf (info.get_symlink_target ());
+                } catch (Error e) {
+                    warning ("Could not get info for %s", file.get_uri ());
+                    primary = _("Replace the target of “%s”?").printf (display_name);
+                }
+
+                secondary = _("Replacing the target file for this link will overwrite its current contents.");
+            }
+
+            var replace_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+                primary, secondary, "dialog-warning", Gtk.ButtonsType.CANCEL
+            ) {
+                modal = true,
+                transient_for = parent
+            };
+
+            var replace_button = replace_dialog.add_button ("Replace", Gtk.ResponseType.YES);
+            replace_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+
+            var response_ok = false;
+            replace_dialog.response.connect ((id) => {
+                response_ok = id == Gtk.ResponseType.YES;
+                check_overwrite_uri.callback ();
+            });
+
+            replace_dialog.present ();
+            yield;
+            replace_dialog.destroy ();
+            return response_ok;
+        } else {
+            return true;
+        }
+    }
```

Comment:
Thinking more about this function, i prefer it begin sync and returning the dialog instead of the dialog response, this way we can handle the `response()` signal directly in the FileChooser callback.

```suggestion
    private Gtk.Dialog create_overwrite_dialog (Gtk.Window parent, GLib.File file) {
        unowned var primary = _("Replace “%s”?");
        unowned var secondary = _("Replacing this file will overwrite its current contents");
        var display_name = file.get_basename ();
        
        if (file.query_file_type (FileQueryInfoFlags.NOFOLLOW_SYMLINKS) == FileType.SYMBOLIC_LINK) {
            try {
                var info = file.query_info (FileAttribute.STANDARD_SYMLINK_TARGET, FileQueryInfoFlags.NONE);
                display_name = info.get_symlink_target ();
            } catch (Error e) {
                warning ("Could not get info for %s", file.get_uri ());
                primary = _("Replace the target of “%s”?");
            }

            secondary = _("Replacing the target file for this link will overwrite its current contents.");
        }

        var replace_dialog = new Granite.MessageDialog.with_image_from_icon_name (
                primary.printf (display_name), secondary, "dialog-warning", Gtk.ButtonsType.CANCEL
            ) {
                modal = true,
                transient_for = parent
            };

        var replace_button = replace_dialog.add_button ("Replace", Gtk.ResponseType.YES);
        replace_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
        
        return replace_dialog;
    }
```